### PR TITLE
New accounting check tool

### PR DESCRIPTION
**Pull request by enolfc**
# Summary

- **Add a site accounting tool**
- **Renamed cli to vm_monitor_cli to avoid conflicts**
- **Add new depedencies and tool**

---

**Related issue :**

### CODE REVIEW

**Comment by enolfc** [comment]
It's working but the linter will scream now. Also needs a bit of cleaner output

**Comment by enolfc** [comment]
I added AppDB checks as well for the SLAs, this makes the tool more integrated with the rest of the repo

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    appdb_vos = set(appdb.get_vo_for_site(site))
+    if site not in site_slas:
+        click.echo(f"[I] {site} is not present in any SLA")
+    else:
+        for sla_name, sla in site_slas[site].items():
+            sla_vos = sla_vos.union(sla["vos"])
+            accounted_vos = sla["vos"].intersection(acct.site_vos(site))
+            if accounted_vos:
+                click.echo(
+                    f"[OK] SITE {site} has accouting info for SLA {sla_name} ({accounted_vos})"
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
                    f"[OK] SITE {site} has accounting info for SLA {sla_name} ({accounted_vos})"
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                )
+            else:
+                click.echo(
+                    f"[ERR] SITE {site} has no accouting info for SLA {sla_name}"
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
                    f"[ERR] SITE {site} has no accounting info for SLA {sla_name}"
```

File: README.md [hunk 3]
```diff
@@ -193,7 +201,33 @@ Getting VMs information  [####################################]  100%
 [-] WARNING: Less than 3 security groups per instance
 ```
 
+## fedcloud-sla-monitor
+
+`fedcloud-sla-monitor` checks the configuration of sites supporting SLAs. It
+compares the reported usage in the accoutnting portal and the information
```
Comment (sebastian-luna-valero) [inline README.md]:
```suggestion
compares the reported usage in the accounting portal and the information
```

**Review by sebastian-luna-valero — COMMENTED** [review]
Thanks!

Minor review until I get my x509 certificate to do a full test

File: fedcloud_vm_monitoring/sla_monitor_cli.py (hunk 1 above)
Comment (enolfc) [reply fedcloud_vm_monitoring/sla_monitor_cli.py]: I don't know how to write accounting :) 
It reminds me of a writing on a wall near home:
`Los disléxicos también somos persianas`

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+import click
+import yaml
+from fedcloud_vm_monitoring.accounting import Accounting
+from fedcloud_vm_monitoring.appdb import AppDB
+from fedcloud_vm_monitoring.goc import GOCDB
+
+
+def check_site_slas(site, site_slas, goc, acct, appdb):
+    click.echo(f"[-] Checking site {site}")
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
    click.secho(f"[-] Checking site {site}", fg="blue", bold=True)
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    appdb_vos = set(appdb.get_vo_for_site(site))
+    if site not in site_slas:
+        click.echo(f"[I] {site} is not present in any SLA")
+    else:
+        for sla_name, sla in site_slas[site].items():
+            sla_vos = sla_vos.union(sla["vos"])
+            accounted_vos = sla["vos"].intersection(acct.site_vos(site))
+            if accounted_vos:
+                click.echo(
+                    f"[OK] SITE {site} has accounting info for SLA {sla_name} ({accounted_vos})"
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
                    f"[OK] {site} has accounting info for SLA {sla_name} ({accounted_vos})"
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                )
+            else:
+                click.echo(
+                    f"[ERR] SITE {site} has no accounting info for SLA {sla_name}"
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
                    f"[ERR] {site} has no accounting info for SLA {sla_name}"
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                )
+            info_vos = sla["vos"].intersection(appdb_vos)
+            if info_vos:
+                click.echo(
+                    f"[OK] SITE {site} has configured {info_vos} for SLA {sla_name}"
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
                    f"[OK] {site} has configured {info_vos} for SLA {sla_name}"
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                )
+            else:
+                click.echo(f"[ERR] SITE {site} has no configured VO for SLA {sla_name}")
+    click.echo("[-] Checking aditional VOs")
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
                click.echo(f"[ERR] {site} has no configured VO for SLA {sla_name}")
    click.secho(f"[-] Checking aditional VOs at {site}", fg="yellow", bold=True)
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    # Now check which VOs are being reported without a SLA
+    if not sla_vos:
+        sla_vos = goc.sla_vos
+    non_sla_vos = acct.site_vos(site) - sla_vos.union(set(["ops"]))
+    if non_sla_vos:
+        click.echo(
+            f"[W] Site {site} has accounting for VOs {non_sla_vos} but non covered by SLA"
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
            f"[W] {site} has accounting for VOs {non_sla_vos} but not covered by SLA"
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        )
+    if "ops" not in acct.site_vos(site):
+        click.echo(f"[W] SITE {site} has accounting for ops")
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
        click.echo(f"[W] {site} has accounting for ops")
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    non_sla_appdb_vos = appdb_vos - sla_vos.union(set(["ops"]))
+    if non_sla_vos:
+        click.echo(
+            f"[W] Site {site} has VOs {non_sla_appdb_vos} configured but non covered by SLA"
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
            f"[W] {site} has VOs {non_sla_appdb_vos} configured but not covered by SLA"
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        )
+    if "ops" not in appdb_vos:
+        click.echo(f"[W] SITE {site} has no configuration for ops")
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
        click.echo(f"[W] {site} has no configuration for ops")
    click.echo()
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        vo_map_src = importlib.resources.read_text(
+            "fedcloud_vm_monitoring.data", "vos.yaml"
+        )
+    vo_map = yaml.load(vo_map_src, Loader=yaml.SafeLoader)
+    acct = Accounting()
+    goc = GOCDB()
+    appdb = AppDB()
+    slas = goc.get_sites_slas(user_cert, vo_map)
+
+    print(slas)
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
    #print(slas)
```

**Review by sebastian-luna-valero — COMMENTED** [review]
Suggestions for prettier output

**Comment by sebastian-luna-valero** [comment]
Additionally, I am troubleshooting something else so please let me get back to you on that later

File: fedcloud_vm_monitoring/sla_monitor_cli.py (hunk 13 above)
Comment (enolfc) [reply fedcloud_vm_monitoring/sla_monitor_cli.py]: I think it's better to just remove the line

Comment (enolfc) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+def check_site_slas(site, site_slas, goc, acct, appdb):
+    click.secho(f"[-] Checking site {site}", fg="blue", bold=True)
+    sla_vos = set()
+    appdb_vos = set(appdb.get_vo_for_site(site))
+    if site not in site_slas:
+        click.echo(f"[I] {site} is not present in any SLA")
+    else:
+        for sla_name, sla in site_slas[site].items():
+            sla_vos = sla_vos.union(sla["vos"])
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py:21]:
```suggestion
            click.echo(f"Information for SLA {sla_name}")
            sla_vos = sla_vos.union(sla["vos"])
```

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+                click.echo(
+                    f"[OK] {site} has accounting info for SLA {sla_name} ({accounted_vos})"
+                )
+            else:
+                click.echo(f"[ERR] {site} has no accounting info for SLA {sla_name}")
+            info_vos = sla["vos"].intersection(appdb_vos)
+            if info_vos:
+                click.echo(f"[OK] {site} has configured {info_vos} for SLA {sla_name}")
+            else:
+                click.echo(f"[ERR] {site} has no configured VO for SLA {sla_name}")
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py:33]:
```suggestion
                click.echo(f"[ERR] {site} has no configured VO for SLA {sla_name}")
            click.echo()
```

**Review by sebastian-luna-valero — COMMENTED** [review]
Further suggestions to improve output

File: fedcloud_vm_monitoring/goc.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            endpoints = group.get("SERVICE_ENDPOINT", [])
+            if not isinstance(endpoints, list):
+                endpoints = [endpoints]
+            for endpoint in endpoints:
+                svc = self.get_endpoint_site(endpoint)
+                if svc:
+                    site = svc["SITENAME"]
+                    site_info = sites.get("site", dict())
+                    site_info[sla_name] = {"vos": set(vos or [])}
+                    sites[site] = site_info
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/goc.py]:
```suggestion
                    if site in sites:
                        sites[site].update(site_info)
                    else:
                        sites[site] = site_info
```

**Review by sebastian-luna-valero — CHANGES_REQUESTED** [review]
Update existing info, otherwise it will be overwritten

File: fedcloud_vm_monitoring/sla_monitor_cli.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    # Now check which VOs are being reported without a SLA
+    if not sla_vos:
+        sla_vos = goc.sla_vos
+    non_sla_vos = acct.site_vos(site) - sla_vos.union(set(["ops"]))
+    if non_sla_vos:
+        click.echo(
+            f"[W] {site} has accounting for VOs {non_sla_vos} but not covered by SLA"
+        )
+    if "ops" not in acct.site_vos(site):
+        click.echo(f"[W] {site} has accounting for ops")
```
Comment (sebastian-luna-valero) [inline fedcloud_vm_monitoring/sla_monitor_cli.py]:
```suggestion
        click.echo(f"[W] {site} has no accounting for ops")
```

**Review by sebastian-luna-valero — CHANGES_REQUESTED** [review]
Missing `no`

**Review by sebastian-luna-valero — APPROVED** [review]
Many thanks!

Let's give it a try. I have a couple of additional ideas but I will open separate PRs for them.

**Pull request merged**